cli -netinfo peer connections dashboard updates 🎄 ✨

jonatack opened the pull request:
Merry Bitcoin Christmas! Ho ho ho 🎄 ✨

This PR updates `-netinfo` to:
- use the getpeerinfo `connection_type` field (and no longer use getpeerinfo `relaytxes` for block-relay detection)
- display manual peers count, if any, in the outbound row
- display the block relay counts in the outbound row only
- display high-bandwidth BIP152 compact block relay peers (`hb` column, to `.` and from `*`)
- add support for displaying I2P network peers, if any are present

Testing and review welcome! How to test:

- to run the full live dashboard (on Linux): `$ watch --interval 1 --no-title ./src/bitcoin-cli -netinfo 4`
- to run the full dashboard: ``$ ./src/bitcoin-cli -netinfo 4``
- to see the help: `$ ./src/bitcoin-cli -netinfo help`
- to see the help summary: `$ ./src/bitcoin-cli -help | grep -A4 netinfo`

jonatack committed:
netinfo: add i2p network

the i2p peer counts column is displayed iff the node is connected
to at least one i2p peer, so this doesn't add clutter for users
who are not running an i2p service

jonatack committed:
netinfo: update to use peer connection types

jonatack committed:
netinfo: add ConnectionTypeForNetinfo member helper function

jonatack committed:
netinfo: display manual peers count

jonatack committed:
netinfo: add bip152 high-bandwidth to/from fields

jonatack commented:
Sample output:
```
$ ./src/bitcoin-cli -netinfo 1
Bitcoin Core v21.99.0-ebfe51a48d-dirty - 70016/Satoshi:21.99.0/

Peer connections sorted by direction and min ping
<->   type   net  mping   ping send recv  txn  blk hb age  asmap  id 
out  block  ipv4     89     92   13   12        36 *  366  34779 754 
out   full  ipv4    135    138    2    3    0   72 *  368  27257 747 
out   full  ipv4    167    176    2    1    0         323    855 776 
out manual onion    188    210    0    0    0         366        757 
out manual onion    206    344    0    2    0         364        764 
out   full onion    219    354    0    0    0         286        785 
out   full  ipv4    223    240    1   11    0         367   7922 749 
out   full  ipv4    241    243    2    7    0         367  37611 748 
out manual onion    257    345    0    0    0         364        765 
out   full onion    277    608  110   19              355        767 
out   full  ipv4    284    290    1    2    0         367   9318 751 
out   full  ipv4    296    298    1    1    0  323 *  368 133938 746 
out manual onion    297    523    5    4    0         364        763 
out  block  ipv4    305    311   44   44              367  38195 753 
out manual onion    321   2086    2    2    0         294        784 
out manual onion    363    656    0    0   19         366        756 
out manual onion    498    864    1   11    0         365        760 
out manual onion    724   1589    2    7   11         365        759 
out  block onion   2623   2623    2    2                0        848 
                     ms     ms  sec  sec  min  min    min

        ipv4    ipv6   onion   total   block  manual
in         0       0       0       0       0       0
out        8       0      11      19       3       8
total      8       0      11      19       3       8

Local addresses
********************************************************.onion     port   8333    score      4
```

jonatack commented:
Sample of (partial) output:

![Screenshot from 2020-12-24 21-08-02](https://user-images.githubusercontent.com/2415484/103105212-3e204000-4624-11eb-8151-02a86cf4f626.jpg)

jonatack commented:
Sample (I believe it is demonstrating a 3rd outbound-block-relay connection to sync headers, introduced in #19858)

![Screenshot from 2020-12-26 13-05-22](https://user-images.githubusercontent.com/2415484/103151132-45099880-4773-11eb-9626-663e92f46e37.png)

sipa commented:
Concept ACK

michaelfolkson commented:
Concept ACK. Is this ready to move out of draft? Would perhaps get more review if out of draft?

practicalswift commented:
Concept ACK

`-netinfo` is great! :)

jonatack commented:
Moved out of draft and ready for review with updated PR description and "how to test" suggestions. The last commit that adds displaying BIP152 high-bandwidth peers can't be backported to rc5, but the other commits, including the user help, could be if desired.

jonasschnelli commented:
Neat!
Concept ACK

glozow reviewed (commented):
Concept ACK, I tried it out - very nice :)

An idea - you could omit the `txn` column during IBD since the node isn't validating (and shouldn't be receiving) txns at all.

The second table should be mentioned in `help` imo, and in general just seems really weird to me. We have a column _after_ the "overall totals" and it has a slot for inbound block-relay-only, which doesn't exist afaik? Seems like it would make more sense for block/manual to be rows instead or have a separate table or something.

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -349,6 +352,72 @@ class NetinfoRequestHandler : public BaseRequestHandler
     }
+    std::string ConnectionTypeForNetinfo(const std::string& conn_type) const
+    {
+        if (conn_type == "outbound-full-relay" || conn_type == "inbound") return "full";
+        if (conn_type == "block-relay-only") return "block";
+        if (conn_type == "manual" || conn_type == "feeler") return conn_type;
+        if (conn_type == "addr-fetch") return "addr";
```
Yes, if an integral or enum connection type was exposed, but getpeerinfo provides it as a string, so this is the best we can do, I think. 

https://en.cppreference.com/w/cpp/language/switch

jonatack replied:
(Client side, I think we have to make do data-wise with what the server side rpc provides. We can't hook in here to CNodeStats; it would be a layer violation).

laanwj replied:
Yea it's too bad that C++ doesn't have case pattern matching on strings, otherwise that'd be a lot cleaner.
(this applies to some other places as well such as the P2P message dispatch)

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -349,6 +352,72 @@ class NetinfoRequestHandler : public BaseRequestHandler
+            "         \"addr\"   - address fetch; short-lived connection for requesting addresses\n"
+            "net      Network the peer connected through (\"ipv4\", \"ipv6\", \"onion\", \"i2p\", or \"cjdns\")\n"
+            "mping    Minimum observed ping time in milliseconds (ms)\n"
+            "ping     Ping time in milliseconds (ms)\n"
```
Good idea!

jonatack replied:
Thanks @glozow, changed to "Last observed ping time." Should go well with its neighbor, "Minimum observed ping time."

jonatack commented on src/bitcoin-cli.cpp:
```diff
@@ -331,7 +333,8 @@ class NetinfoRequestHandler : public BaseRequestHandler
         int id;
         int mapped_as;
         int version;
-        bool is_block_relay;
```
Hm, interesting. Restoring this (`fRelayTxes`) is trivial if needed. I'd prefer to remove any unused struct members here, more so as this struct is a mini-hotspot in the peers loop and -netinfo should be callable at somewhat high frequencies.

jonatack replied:
Yes, dropped the commit in order to keep `is_block_relay` and reworked the connection type logic to distinguish between inbound full relay (based on `fRelayTxes` like in `AttemptToEvictConnection`) and outbound block relay (based on `m_conn_type`).

jonatack commented:
Thanks for your ideas @glozow!

> The second table should be mentioned in `help` imo, and in general just seems really weird to me. We have a column _after_ the "overall totals" and it has a slot for inbound block-relay-only, which doesn't exist afaik? Seems like it would make more sense for block/manual to be rows instead or have a separate table or something.

This is valuable, as I don't have a fresh perspective. I'd like to keep the connection counts in as compact a format as we can, but any ideas welcome. How about this? (If inbound block-relay connections happen then the block column could go back to a full one. Note that the current -netinfo is already in the 0.21 release, with a full block-relay column but not a manual column.)

![Screenshot from 2020-12-29 22-21-12](https://user-images.githubusercontent.com/2415484/103314927-0fd3a380-4a1c-11eb-971d-44e82258cf91.png)

jonatack commented:
Updated:
- Rewrote the help, adding suggestions by @glozow and improved formatting
- Display the block relay and manual counts in the outbound row only

jonatack commented:
<details><summary>Updated help doc</summary><p>

```
$ ./src/bitcoin-cli -netinfo help
-netinfo level "help" 

Returns a network peer connections dashboard with information from the remote server.
Under the hood, -netinfo fetches the data by calling getpeerinfo and getnetworkinfo.
An optional integer argument from 0 to 4 can be passed for different peers listings.
Pass "help" to see this detailed help documentation.
If more than one argument is passed, only the first one is read and parsed.
Suggestion: use with the Linux watch command for a live dashboard; see example below.

Arguments:
1. level (integer 0-4, optional)  Specify the info level of the peers dashboard (default 0):
                                  0 - Connection counts and local addresses
                                  1 - Like 0 but with a peers listing (without address or version columns)
                                  2 - Like 1 but with an address column
                                  3 - Like 1 but with a version column
                                  4 - Like 1 but with both address and version columns
2. help (string "help", optional) Print this help documentation instead of the dashboard.

Result:

* The peers listing in levels 1-4 displays all of the peers sorted by direction and minimum ping time:

  Column   Description
  ------   -----------
  <->      Direction
           "in"  - inbound connections are those initiated by the peer
           "out" - outbound connections are those initiated by us
  type     Type of peer connection
           "full"   - full relay, the default
           "block"  - block relay; like full relay but does not relay transactions or addresses
           "manual" - peer we manually added using RPC addnode or the -addnode/-connect config options
           "feeler" - short-lived connection for testing addresses
           "addr"   - address fetch; short-lived connection for requesting addresses
  net      Network the peer connected through ("ipv4", "ipv6", "onion", "i2p", or "cjdns")
  mping    Minimum observed ping time, in milliseconds (ms)
  ping     Last observed ping time, in milliseconds (ms)
  send     Time since last message sent to the peer, in seconds
  recv     Time since last message received from the peer, in seconds
  txn      Time since last novel transaction received from the peer and accepted into our mempool, in minutes
  blk      Time since last novel block passing initial validity checks received from the peer, in minutes
  hb       High-bandwidth BIP152 compact block relay
           "." (to)   - we selected the peer as a high-bandwidth peer
           "*" (from) - the peer selected us as a high-bandwidth peer
  age      Duration of connection to the peer, in minutes
  asmap    Mapped AS (Autonomous System) number in the BGP route to the peer, used for diversifying
           peer selection (only displayed if the -asmap config option is set)
  id       Peer index, in increasing order of peer connections since node startup
  address  IP address and port of the peer
  version  Peer version and subversion concatenated, e.g. "70016/Satoshi:21.0.0/"

* The connection counts table displays the number of peers by direction, network, and the totals
  for each, as well as two special outbound columns for block relay peers and manual peers.

* The local addresses table lists each local address broadcast by the node, the port, and the score.

Examples:

Connection counts and local addresses only
> bitcoin-cli -netinfo

Compact peers listing
> bitcoin-cli -netinfo 1

Full dashboard
> bitcoin-cli -netinfo 4

Full live dashboard, adjust --interval or --no-title as needed (Linux)
> watch --interval 1 --no-title ./src/bitcoin-cli -netinfo 4

See this help
> bitcoin-cli -netinfo help

```
</p></details>

theStack commented:
Very nice that Santa Claus also came to present a gift to Bitcoin Core ;) 🎅 
Concept ACK, plan to review this within the next days.

jonatack commented:
Rebased on top of #20829.

felipsoarez reviewed (commented):
Concept ACK, not tested

Good job, more information for better adaptations `-netinfo` is great

jonatack commented:
Rebased after the merge of #20829. The diff is now much smaller and this should be ready for final review.

jonatack commented:
Dropped two commits

- 23e4faa5 "netinfo: remove now-unused is_block_relay"
- 97127a870 "netinfo: only display outbound-block-relay peers count"

and reworked the connection type logic to distinguish between

- `inbound-block-relay` peers based on (`fInbound` and) `fRelayTxes` like in `AttemptToEvictConnection`
- `outbound-block-relay` peers based on `m_conn_type`

and count both in the block relay totals.

jonatack commented:
Added a commit to support displaying connections to peers via the I2P network. An `i2p` peer counts column is displayed if the user is connected to any I2P peers. Otherwise there is no user-facing change, and this change is compatible with current master.

![Screenshot from 2021-01-17](https://user-images.githubusercontent.com/2415484/104956769-b7f9de00-59c4-11eb-824c-3f0cc183f65b.png)

jonatack committed:
netinfo: display only outbound block relay counts

jonatack commented:
Updated connection type handling to display inbounds without a "full" or "block" type description from fRelayTxes.

Will give this a few days and if still no ACKs will close in favor of proposing the changes individually in separate pulls.

michaelfolkson commented:
You haven't started splitting this into separate pulls have you @jonatack? If no I will review this tomorrow. Otherwise I will wait for the separate pulls.

jonatack commented:
Thanks 🙏 leaving as-is for now, one merge is easier than several 🐇 🐢

laanwj commented:
Code review and shallowly tested ACK 9628f9953d81c380c780046ef30a2f61c96d4966

I am a bit divided on adding I2P support here before #20685 is merged, but also don't see it as a problem that the client can already handle this.

MarcoFalke commented on src/bitcoin-cli.cpp:
```diff
@@ -489,22 +513,23 @@ class NetinfoRequestHandler : public BaseRequestHandler
+                    "%3s %6s %5s%7s%7s%5s%5s%5s%5s  %2s %*s%*i %*s %-*s%s\n",
                     peer.is_outbound ? "out" : "in",
-                    peer.is_block_relay ? "block" : "full",
+                    ConnectionTypeForNetinfo(peer.conn_type),
```
In your screenshot an i2p inbound peer will print "full", which is short for "outbound-full-relay". Is this something wrong with i2p or with this pull?

jonatack replied:
The screenshot in the PR description was not the latest version. Removed it.

jonatack replied:
The inbound handling was updated after the i2p screenshot.

jonatack commented:
Updated the last commit to only display block relay peer counts in the outbound row.

`git diff 9628f99 a2dd891`

jonatack commented:
Updated the last commit to only display block relay peer counts in the *outbound* row, and simplified it and commit *netinfo: display manual peers count* by persisting the counts in a member variable instead of in the `m_counts` 2D array.

@laanwj Thanks for reviewing! I apologize for pushing again after. I think this is finally done.

michaelfolkson commented:
ACK 747cb5b9949f80b3b4516f382a0ce80e41f3f5a6

I really like it (obviously) and I've only done limited testing (no Tor or I2P connections)

> display high-bandwidth BIP152 compact block relay peers (hb column, to . and from *)

Maybe it is just me but I feel as if I'm going to have to look at the help every time to remember what `.` and `*` represents. Presumably there were no better options than `.` and `*`? Obviously `in` and `out` are already taken and `to` and `from` took up too much space?

My understanding is a little hazy of BIP 152 selection process too so maybe I personally just wouldn't care about this. I can look it up though!

jonatack commented:
Thanks @michaelfolkson for reviewing! The idea is `.` for an outgoing signal ("to"; we selected the peer) and `*` for an inbound signal ("from"; the peer selected us). Kind of like a train with a bright headlight and a muted tail light. Though if you see them as starships instead I suppose an argument could be made for the inverse :smile: 
 
![Screenshot from 2020-12-29](https://user-images.githubusercontent.com/2415484/106957967-9a50a680-6730-11eb-85c7-a4f91263e86f.png)

jonasschnelli commented:
Should this also be compatible with `-testnet` (and eventually `-regtest)?

jonasschnelli reviewed (approved):
Tested ACK 747cb5b9949f80b3b4516f382a0ce80e41f3f5a6 - works nicely. Great that this PR only changes bitcoin-cli.
